### PR TITLE
feat(lambda): collecting status code from result

### PR DESCRIPTION
**Pull request by galbash**

### CODE REVIEW

File: epsagon/event.py [hunk 1]
```diff
@@ -105,3 +107,5 @@ def set_exception(self, exception, traceback_data):
         self.exception['message'] = str(exception)
         self.exception['traceback'] = traceback_data
         self.exception['time'] = time.time()
+        if not handled:
```
Comment (ranrib) [inline epsagon/event.py]: Wouldn't it be a bit more clear to call it "manually_set" instead of handled/unhandled?

Comment (ronyis) [reply epsagon/event.py]: I think that 'handled' it's good. It can be 'handled' and not manually set - for example, if we will extract errors from calling `logging.error`

File: epsagon/event.py [hunk 2]
```diff
@@ -105,3 +107,5 @@ def set_exception(self, exception, traceback_data):
...
+            self.exception.setdefault('additional_data', {})['unhandled'] = True
```
Comment (ronyis) [inline epsagon/event.py]: I think it's more straight forward to save `{'handled': True}` so that the trace will be not changed in the most common event of unhandled exception.

Comment (ronyis) [inline epsagon/event.py]: Are you sure that `additional_data` is needed?

File: epsagon/events/botocore.py [hunk 3]
```diff
@@ -64,14 +64,15 @@ def __init__(self, wrapped, instance, args, kwargs, start_time, response,
...
-        :param exception: Exception object
-        :param traceback_data: traceback string
-        :return: None
+        see {Event.set_exception}
         """
-        super(BotocoreEvent, self).set_exception(exception, traceback_data)
+        super(BotocoreEvent, self).set_exception(
+            exception,
+            traceback_data,
+            handled
```
Comment (ronyis) [inline epsagon/events/botocore.py]: Keyword args should be called with the keyword.

File: epsagon/wrappers/aws_lambda.py [hunk 4]
```diff
@@ -94,10 +108,11 @@ def _lambda_wrapper(*args, **kwargs):
             return result
         # pylint: disable=W0703
         except Exception as exception:
-            runner.set_exception(exception, traceback.format_exc())
+            runner.set_exception(exception, traceback.format_exc(), False)
```
Comment (ronyis) [inline epsagon/wrappers/aws_lambda.py]: Call with keyword (in all places)

File: epsagon/wrappers/aws_lambda.py [hunk 5]
```diff
@@ -22,6 +23,19 @@
...
 
+def _add_status_code(runner, return_value):
+    """
+    Tries to extract the status code from the return value and add it
+    as a metadata field
+    :param runner: Runner event to update
+    :param return_value: The return value to extract from
+    """
+    if isinstance(return_value, collections.Mapping):
+        status_code = return_value.get('statusCode', None)
```
Comment (ronyis) [inline epsagon/wrappers/aws_lambda.py]: `None` is not needed

File: epsagon/wrappers/aws_lambda.py [hunk 6]
```diff
@@ -94,10 +108,11 @@ def _lambda_wrapper(*args, **kwargs):
             return result
         # pylint: disable=W0703
         except Exception as exception:
-            runner.set_exception(exception, traceback.format_exc())
+            runner.set_exception(exception, traceback.format_exc(), False)
             raise
         finally:
             try:
+                _add_status_code(runner, result)
```
Comment (ronyis) [inline epsagon/wrappers/aws_lambda.py]: It would be better if you collect it only for API-GW / load balancer trigger, and I think you should save it in the trigger.
(Now you could mistakenly collect this data when it's `metadata_only=True`)

File: epsagon/wrappers/aws_lambda.py [hunk 7]
```diff
@@ -206,10 +221,11 @@ def _lambda_wrapper(*args, **kwargs):
             return result
         # pylint: disable=W0703
         except Exception as exception:
-            runner.set_exception(exception, traceback.format_exc())
+            runner.set_exception(exception, traceback.format_exc(), False)
```
Comment (ronyis) [inline epsagon/wrappers/aws_lambda.py]: Code duplication is so fun! 🎉

File: epsagon/wrappers/aws_lambda.py [hunk 8]
```diff
@@ -206,10 +221,11 @@ def _lambda_wrapper(*args, **kwargs):
...
             raise
         finally:
             try:
+                _add_status_code(runner, result)
```
Comment (ronyis) [inline epsagon/wrappers/aws_lambda.py]: Step functions can be triggered by HTTP event?

File: tests/test_trace.py [hunk 9]
```diff
@@ -861,7 +861,7 @@ def test_event_with_datetime(wrapped_post):
 def test_send_on_error_only_off_with_error(wrapped_post):
     trace = trace_factory.get_or_create_trace()
     trace.token = 'a'
-    trace.runner = mock.MagicMock()
+    trace.runner = RunnerEventMock()
```
Comment (ronyis) [inline tests/test_trace.py]: This is not really related to your PR or I missed something?

File: epsagon/wrappers/aws_lambda.py (hunk 6 above)
Comment (galbash) [reply epsagon/wrappers/aws_lambda.py]: we agreed to allow it

File: epsagon/wrappers/aws_lambda.py (hunk 7 above)
Comment (galbash) [reply epsagon/wrappers/aws_lambda.py]: :)

Comment (galbash) [reply epsagon/wrappers/aws_lambda.py]: this is why node is better 😎

File: epsagon/wrappers/aws_lambda.py (hunk 8 above)
Comment (galbash) [reply epsagon/wrappers/aws_lambda.py]: actually i think it can :)

File: tests/test_trace.py (hunk 9 above)
Comment (galbash) [reply tests/test_trace.py]: fixing the random crashes in the tests

Comment (ronyis) [reply tests/test_trace.py]: 👑
